File: August2022/C++/RansomNote.cpp
// https://leetcode.com/problems/ransom-note/
class Solution {
public:
    // TC: O(magazine.length() + ransomNote.length())
    // SC: O(26) ~ O(1)
    bool canConstruct(string ransomNote, string magazine) {
        vector<int> letters(26, 0);
        
        for (int i = 0; i < magazine.length(); i++) {
            char c = magazine[i];
            letters[c - 'a']++;
        }
        for (int i = 0; i < ransomNote.length(); i++) {
            char c = ransomNote[i];
            if (letters[c - 'a'] == 0) {
                return false;
            }
            letters[c - 'a']--;
        }
        
        return true;
    }
};

File: August2022/Java/RansomNote.java

// @saorav21994
// TC : O(max(magazine,ransomNote))
// SC : O(26) = O(1)

class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        int [] hash = new int[26];
        for (char ch : magazine.toCharArray()) {
            hash[(int)ch-97] += 1;
        }
        for (char ch : ransomNote.toCharArray()) {
            hash[(int)ch-97] -= 1;
            if (hash[(int)ch-97] < 0)
                return false;
        }
        return true;
        
    }
}
